python/tests/catalog/test_base.py

```diff
@@ -0,0 +1,444 @@
+
+    def replace_table(
+        self,
+        *,
+        namespace: Tuple[str, ...],
+        name: str,
+        schema: Schema,
+        location: Optional[str] = None,
+        partition_spec: Optional[PartitionSpec] = None,
+        properties: Optional[Dict[str, str]] = None,
+    ) -> Table:
+
+        try:
+            table = self.__tables.pop((namespace, name))
+        except KeyError:
+            raise TableNotFoundError(f"Table {name} not found in the catalog")
+
+        replaced_table = InMemoryTable(
+            namespace=namespace if namespace else table.namespace,
+            name=name if name else table.name,
+            schema=schema if schema else table.schema,
+            location=location if location else table.location,
+            partition_spec=partition_spec if partition_spec else table.partition_spec,
+            properties={**table.properties, **properties},
+        )
+        self.__tables[(replaced_table.namespace, replaced_table.name)] = replaced_table
+        return replaced_table
+
+    def create_namespace(self, namespace: Tuple[str, ...], properties: Optional[Dict[str, str]] = None) -> None:
+        if namespace in self.__namespaces:
+            raise AlreadyExistsError(f"Namespace {namespace} already exists")
+        else:
+            self.__namespaces[namespace] = properties if properties else {}
+
+    def drop_namespace(self, namespace: Tuple[str, ...]) -> None:
+        if [table_name_tuple for table_name_tuple in self.__tables.keys() if namespace in table_name_tuple]:
+            raise NamespaceNotEmptyError(f"Namespace {namespace} not empty")
+        try:
+            self.__namespaces.pop(namespace)
+        except KeyError:
+            raise NamespaceNotFoundError(f"Namespace {namespace} not found in the catalog")
+
+    def list_tables(self, namespace: Optional[Tuple[str, ...]] = None) -> List[Tuple[Tuple[str, ...], str]]:
+        if namespace:
+            list_tables = [table_name_tuple for table_name_tuple in self.__tables.keys() if namespace in table_name_tuple]
+        else:
+            list_tables = list(self.__tables.keys())
+
+        # Casting to make mypy happy
+        return cast(List[Tuple[Tuple[str, ...], str]], list_tables)
+
+    def list_namespaces(self) -> List[Tuple[str, ...]]:
+        return list(self.__namespaces.keys())
+
+    def load_namespace_metadata(self, namespace: Tuple[str, ...]) -> Dict[str, str]:
+        try:
+            return self.__namespaces[namespace]
+        except KeyError:
+            raise NamespaceNotFoundError(f"Namespace {namespace} not found in the catalog")
+
+    def set_namespace_metadata(self, namespace: Tuple[str, ...], metadata: Dict[str, str]) -> None:
+        if namespace in self.__namespaces:
+            self.__namespaces[namespace] = metadata
+        else:
+            raise NamespaceNotFoundError(f"Namespace {namespace} not found in the catalog")
+
+
+@pytest.fixture
+def catalog() -> InMemoryCatalog:
```

Comment:
We may want to use this fixture in other test files so I would suggest moving it to `conftest.py`.